### ISSUE TITLE
NextcloudApiNotRespondingException

### ISSUE THREAD

**Issue by David-Development** [nextcloud/Android-SingleSignOn#162]
By now two apps are having issues with the `NextcloudApiNotRespondingException` Bug.

- https://github.com/stefan-niedermann/nextcloud-notes/issues/653
- https://github.com/nextcloud/news-android/issues/769

A few days ago I was experiencing the same issue on my device (I think right after I updated the news app via the play store). Several restarts of the app did not help. Only setting up the account again solved the problem for me. Still not sure if the Nextcloud Files App has something to do with it or if the problem is on the "client" side. @stefan-niedermann Do you have new insights into why this error happens? Or any ideas if the files app is involved in this? 

Maybe some old Service Connections are reused which are broken? But that wouldn't make sense in terms of the app restarts I performed. Maybe restarting the phone helps..? That would be interesting to find out.

**Comment by stefan-niedermann** [comment]
@tobiasKaminsky just said, this is reproducable when the files app is updated and SSO should recreate the API instance after a 5s timeout 🤷‍♂

**Comment by David-Development** [comment]
I'm wondering if it'll help if we reconnect the api like this:

```java
private void waitForApi() throws NextcloudApiNotRespondingException {
        synchronized (mBound) {
            // If service is not bound yet.. wait
            if(!mBound.get()) {
                Log.v(TAG, "[waitForApi] - api not ready yet.. waiting [" + Thread.currentThread().getName() +  "]");
                try {
                    mBound.wait(10000); // wait up to 10 seconds

                    // If api is still not bound after 10 seconds.. throw an exception
                    if(!mBound.get()) {
                        this.stop();
                        this.start(super.type);
                        mBound.wait(10000); // wait up to 10 seconds
                        // If api is still not bound after 10 seconds.. throw an exception
                        if(!mBound.get()) {
                          throw new NextcloudApiNotRespondingException();
                        }
                    }
                } catch (InterruptedException ex) {
                    Log.e(TAG, "WaitForAPI failed", ex);
                }
            }
        }
    }
```

**Comment by David-Development** [comment]
Nevermind.. needs more than that 😅

**Comment by David-Development** [comment]
Okay, I tried reproducing the issue. I installed the news app and put it into debugging mode. I updated the nextcloud files app using the following command: 

```bash
./adb install ~/Downloads/gplay-release-30100090.apk
```

As I would expect I was able to see the following in the LogCat:

```
2020-01-21 17:41:06.987 5245-5245/de.luhmer.owncloudnewsreader E/com.nextcloud.android.sso.api.AidlNetworkRequest: Nextcloud Single sign-on: ServiceDisconnected
2020-01-21 17:41:06.987 5245-5245/de.luhmer.owncloudnewsreader D/com.nextcloud.android.sso.api.AidlNetworkRequest: reconnecting lost SingleSignOn service connection
2020-01-21 17:41:06.987 5245-5245/de.luhmer.owncloudnewsreader V/com.nextcloud.android.sso.api.NetworkRequest: Nextcloud Single sign-on connect() called [main] Account-Type: [nextcloud]
2020-01-21 17:41:07.334 5245-5245/de.luhmer.owncloudnewsreader V/com.nextcloud.android.sso.api.AidlNetworkRequest: Nextcloud Single sign-on: onServiceConnected [main]
```

So the lost connection (due to the reinstall process) is re-established successfully. So at this point I'm not able to reproduce the issue. I'll need to investigate a little further..

**Comment by teras** [comment]
I've tried connecting with both, the "stable" version of NextCloud Files and the dev version, from the F-Droid repository, as well as the Google Play version. None worked properly. Is there any solution for this yet?

**Comment by David-Development** [comment]
@teras Unfortunately not. I'm on my way back from my vacation and hopefully I'll have some time in the next few days to investigate further. I didn't update my nextcloud files app on the phone yet so I am hoping to be able to reproduce the error and grab some logcat output from it. Like I said before manually reinstalling the files app using `adb` does not help much as everything works as expected during my tests (the api reconnects immediately).

*Call for action*: If anyone wants to help to investigate this issue, it would be very helpful to grab the logcat output (of both the nextcloud files app as well as the client app; e.g. nextcloud news / nextcloud notes) while updating the nextcloud files app trough the app store.

**Comment by David-Development** [comment]
Looks like I was able to fix this issue today. Planning on releasing a new version in the next couple days.

Reference: https://github.com/nextcloud/Android-SingleSignOn/pull/163

**Comment by stefan-niedermann** [comment]
Here we go again :(

SSO-Version `0.5.0-rc3`

https://github.com/stefan-niedermann/nextcloud-notes/issues/752?notification_referrer_id=MDE4Ok5vdGlmaWNhdGlvblRocmVhZDczMjIzOTg2Mzo0NzQxMTk5

**Comment by stefan-niedermann** [comment]
Another one: https://github.com/stefan-niedermann/nextcloud-notes/issues/753#issuecomment-592944692

**Comment by stefan-niedermann** [comment]
Another one with 2 stack traces:

https://github.com/stefan-niedermann/nextcloud-notes/issues/710#issuecomment-593069374

(If i receive another stacktrace i will merge all references to one comment to not flood this thread)

**Comment by FredMa01** [comment]
I am using these two apps :
News 0.9.9.36 (https://f-droid.org/fr/packages/de.luhmer.owncloudnewsreader/)
and 
Notes 2.10.3 (https://f-droid.org/fr/packages/it.niedermann.owncloud.notes/)
with Nextcloud 18.0.1.3
News is ok
Notes is ok one or two synchronisations and after I have issue with "NextcloudApiNotRespondingException Bug"

> 
> App Version: 2.10.3
> App Version Code: 81
> Files App Version Code: 30100190OS Version: 4.9.148(9.1.0.155C432)
> OS API Level: 28
> Device: HWLLD-H
> Model (and Product): LLD-L31 (LLD-L31)com.nextcloud.android.sso.exceptions.NextcloudApiNotRespondingException: L'application Nexcloud Files ne répond pas. Veuillez signaler ce problème.
> at com.nextcloud.android.sso.api.AidlNetworkRequest.waitForApi(AidlNetworkRequest.java:147)
> at com.nextcloud.android.sso.api.AidlNetworkRequest.performAidlNetworkRequestV2(AidlNetworkRequest.java:278)
> at com.nextcloud.android.sso.api.AidlNetworkRequest.performNetworkRequestV2(AidlNetworkRequest.java:165)
> at com.nextcloud.android.sso.api.NextcloudAPI.performNetworkRequestV2(NextcloudAPI.java:148)
> at it.niedermann.owncloud.notes.persistence.NotesClient.requestServer(NotesClient.java:163)
> at it.niedermann.owncloud.notes.persistence.NotesClient.putNote(NotesClient.java:107)
> at it.niedermann.owncloud.notes.persistence.NotesClient.editNote(NotesClient.java:122)
> at it.niedermann.owncloud.notes.persistence.NoteServerSyncHelper$SyncTask.pushLocalChanges(NoteServerSyncHelper.java:347)
> at it.niedermann.owncloud.notes.persistence.NoteServerSyncHelper$SyncTask.doInBackground(NoteServerSyncHelper.java:318)
> at it.niedermann.owncloud.notes.persistence.NoteServerSyncHelper$SyncTask.doInBackground(NoteServerSyncHelper.java:284)
> at android.os.AsyncTask$2.call(AsyncTask.java:334)
> at java.util.concurrent.FutureTask.run(FutureTask.java:266)
> at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:246)
> at java.util.concurrent.ThreadPoolExecutor.processTask(ThreadPoolExecutor.java:1187)
> at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)
> at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
> at java.lang.Thread.run(Thread.java:784)
> com.nextcloud.android.sso.exceptions.NextcloudApiNotRespondingException: L'application Nexcloud Files ne répond pas. Veuillez signaler ce problème.
> at com.nextcloud.android.sso.api.AidlNetworkRequest.waitForApi(AidlNetworkRequest.java:147)
> at com.nextcloud.android.sso.api.AidlNetworkRequest.performAidlNetworkRequestV2(AidlNetworkRequest.java:278)
> at com.nextcloud.android.sso.api.AidlNetworkRequest.performNetworkRequestV2(AidlNetworkRequest.java:165)
> at com.nextcloud.android.sso.api.NextcloudAPI.performNetworkRequestV2(NextcloudAPI.java:148)
> at it.niedermann.owncloud.notes.persistence.NotesClient.requestServer(NotesClient.java:163)
> at it.niedermann.owncloud.notes.persistence.NotesClient.getNotes(NotesClient.java:98)
> at it.niedermann.owncloud.notes.persistence.NoteServerSyncHelper$SyncTask.pullRemoteChanges(NoteServerSyncHelper.java:413)
> at it.niedermann.owncloud.notes.persistence.NoteServerSyncHelper$SyncTask.doInBackground(NoteServerSyncHelper.java:320)
> at it.niedermann.owncloud.notes.persistence.NoteServerSyncHelper$SyncTask.doInBackground(NoteServerSyncHelper.java:284)
> at android.os.AsyncTask$2.call(AsyncTask.java:334)
> at java.util.concurrent.FutureTask.run(FutureTask.java:266)
> at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:246)
> at java.util.concurrent.ThreadPoolExecutor.processTask(ThreadPoolExecutor.java:1187)
> at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)
> at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
> at java.lang.Thread.run(Thread.java:784)

**Comment by PinkFreud** [comment]
I'm seeing this issue as well:

```
App Version: 2.11.3
App Version Code: 2011003
Files App Version Code: 30100190

---

OS Version: 4.14.150-g05fe60fb845b-ab6126650(6156912)
OS API Level: 29
Device: coral
Model (and Product): Pixel 4 XL (coral)

---

com.nextcloud.android.sso.exceptions.NextcloudApiNotRespondingException: Nextcloud files app api is not responding. Please report this issue.
	at com.nextcloud.android.sso.api.AidlNetworkRequest.waitForApi(AidlNetworkRequest.java:147)
	at com.nextcloud.android.sso.api.AidlNetworkRequest.performAidlNetworkRequestV2(AidlNetworkRequest.java:278)
	at com.nextcloud.android.sso.api.AidlNetworkRequest.performNetworkRequestV2(AidlNetworkRequest.java:165)
	at com.nextcloud.android.sso.api.NextcloudAPI.performNetworkRequestV2(NextcloudAPI.java:148)
	at it.niedermann.owncloud.notes.persistence.NotesClient.requestServer(NotesClient.java:162)
	at it.niedermann.owncloud.notes.persistence.NotesClient.getNotes(NotesClient.java:98)
	at it.niedermann.owncloud.notes.persistence.NoteServerSyncHelper$SyncTask.pullRemoteChanges(NoteServerSyncHelper.java:451)
	at it.niedermann.owncloud.notes.persistence.NoteServerSyncHelper$SyncTask.doInBackground(NoteServerSyncHelper.java:358)
	at it.niedermann.owncloud.notes.persistence.NoteServerSyncHelper$SyncTask.doInBackground(NoteServerSyncHelper.java:320)
	at android.os.AsyncTask$3.call(AsyncTask.java:378)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:289)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:919)

```

**Comment by stefan-niedermann** [comment]
And did you follow the steps we mentioned earlier in this thread @PinkFreud ?

**Comment by PinkFreud** [comment]
> And did you follow the steps we mentioned earlier in this thread @PinkFreud ?
> 
> **Edit** Sorry @PinkFreud i was in the wrong issue - Please read starting here: [stefan-niedermann/nextcloud-notes#752 (comment)](https://github.com/stefan-niedermann/nextcloud-notes/issues/752#issuecomment-605430344)
> 
> TL;DR: Clear Storage of both apps, Notes **and** the main Nextcloud android app.

This was almost a month ago (sorry, I seem to have missed your reply, argh!), but I did manage to get it working.  I don't recall what I did now, but I may very well have done that.

**Comment by synhme** [comment]
App Version: 2.11.13
App Version Code: 2011013
App Flavor: play

Files App Version Code: 30110190

---

OS Version: 3.18.19+(1524191364)
OS API Level: 23
Device: M5s
Manufacturer: Meizu
Model (and Product): M5s (Meizu-M5s)

---

com.nextcloud.android.sso.exceptions.NextcloudApiNotRespondingException: ANextcloud files app api is not responding. Please report this issue.
 at com.nextcloud.android.sso.api.AidlNetworkRequest.waitForApi(AidlNetworkRequest.java:147)
 at com.nextcloud.android.sso.api.AidlNetworkRequest.performAidlNetworkRequestV2(AidlNetworkRequest.java:278)
 at com.nextcloud.android.sso.api.AidlNetworkRequest.performNetworkRequestV2(AidlNetworkRequest.java:165)
 at com.nextcloud.android.sso.api.NextcloudAPI.performNetworkRequestV2(NextcloudAPI.java:148)
 at it.niedermann.owncloud.notes.persistence.NotesClient.requestServer(NotesClient.java:163)
 at it.niedermann.owncloud.notes.persistence.NotesClient.getNotes(NotesClient.java:99)
 at it.niedermann.owncloud.notes.persistence.NoteServerSyncHelper$SyncTask.pullRemoteChanges(NoteServerSyncHelper.java:431)
 at it.niedermann.owncloud.notes.persistence.NoteServerSyncHelper$SyncTask.doInBackground(NoteServerSyncHelper.java:345)
 at it.niedermann.owncloud.notes.persistence.NoteServerSyncHelper$SyncTask.doInBackground(NoteServerSyncHelper.java:307)
 at android.os.AsyncTask$2.call(AsyncTask.java:295)
 at java.util.concurrent.FutureTask.run(FutureTask.java:237)
 at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:234)
 at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
 at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
 at java.lang.Thread.run(Thread.java:818)

Tried force stopping, clearing cache, reinstalling both Notes and main Nextcloud apps.

Same thing for the Deck app too.

**Comment by synhme** [comment]
Relevant closed Meizu issue: https://github.com/stefan-niedermann/nextcloud-notes/issues/753

I installed Notes and Deck apps for Android on Meizu M5s, and they both didn't work because of the  _NextcloudApiNotRespondingException_ error.

I fixed the issue by disabling _Smart Background_ feature and allowing all Nextcloud related apps to run in the background.

Previously I tried clearing cache, storage and reinstalling apps multiple times without any positive results.

Logs are the same as above, so I won't provide them.

**Comment by dasbiswajit** [comment]
@David-Development Good Day! I wished to integrate SSO with Nextcloud [bookmark](https://gitlab.com/bisada/OCBookmarks/-/issues/27). the latest code is [here](https://gitlab.com/bisada/OCBookmarks/-/tree/sso_fix). However, this is stopping me from releasing the fix for SSO. This feature is already live in Playstore.   I really need to release a fix for this as soon as possible.

**Comment by stefan-niedermann** [comment]
For documentation @David-Development another one where disabling battery optimization worked as solution: https://github.com/stefan-niedermann/nextcloud-notes/issues/710#issuecomment-651054925

**Comment by grinapo** [comment]
Just reiterating what I said in stefan-niedermann/nextcloud-notes#1312 

For me battery optim or autostart (or basically anything) haven't fixed it. 

But **scratching app storage** did: Notes re-requested the NC account then re-pulled the data and now it works. However I have the suspicion that if I do not use the app for a while (weeks?) it will be stopped by the <s>chinese communist party</s> MIUI 12.5.1 and after that it will not be able to reconnect until another full data clear.
